### ISSUE TITLE
Use without xcode - on purely swift package based projects

### ISSUE THREAD

**Issue by doozMen** [birdrides/mockingbird#164]
## New Issue Checklist

- [x] I updated my Mockingbird framework and CLI to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/birdrides/mockingbird/issues)

## Description
*The code I talk about is in a draft PR #165*
As mocking could be very helpful for projects that do not require you to have an Xcode project I would like to alter the project so I can use it on my server side projects. I'm facing some issues with the way the project is setup.

* loading the dlib is not working as it is missing the `swiftSyntaxParserDylib` variable in main
* just opening the `open Package.swift` will make xcode resolve the dependencies. This works but then I cannot build because of my first bullet
* The instructions say you can add Mockingbird to an existing project Package.swift file. But this results in an error if I follow the instructions. *dependency 'Mockingbird' in target 'UseCaseInjectTests' requires explicit declaration; provide the name of the package dependency with*

To solve this I followed the following steps, unsuccessful so far

1. I followed the contribution guidelines and ran the `make setup-project`
2. Then I `cd sources` to inspect the package
3. I tried to add product `Mockingbird` as a dependency to the commented out test target
4. added the needed files for a test target
5.  declared `swiftSyntaxParserDylib` manually as 
```swift
let file = #file
let dataDlib = file.replacingOccurrences(of: "", with: "") // this is a workaround to be able to build it, not pointing to correct file as this depends on how you build it
let swiftSyntaxParserDylib = Resource(
    encodedData: try Data(contentsOf: URL(fileURLWithPath: dataDlib)).base64EncodedString(),
    fileName: "lib_InternalSwiftSyntaxParser.dylib"
)
```

This fixed the build and I could import `Mockingbird` in my added test target.
6. Then I followed the instructions to [Manual-Setup](https://github.com/birdrides/mockingbird/wiki/Manual-Setup)

But I lack a bit of background to be able to get this working. What I like to do on Package.swift only projects is:

* generate the mocks from targets defined in the Package.swift (by default just all except test targets)
* add some code to a git hook pre-push so it generates the mocks and runs tests before I can push

As this is how I would like to use it without the hassle of having to generate an XCode project I filed this issue. Could you help me get started?

### Framework Bugs

Unable to include the `Mockingbird` framework in a purely swift package project without Xcode.

As this outputs error
```
dependency 'Mockingbird' in target '<#MyTestTarget#>' requires explicit declaration; provide the name of the package dependency with
```

Also adding it with `.product(name: "Mockingbird", package: "mockingbird")` did not do the trick.

## Environment

* Mockingbird CLI version (`0.14.0`)
* Xcode and macOS version (11.6)
* Swift version (`5.2`)
* Installation method (SPM)
* Unit testing framework (XCTest)
* Are you using [supporting source files](https://github.com/birdrides/mockingbird#supporting-source-files)? *Not sure how to, maybe this is the problem that I'm missing something?*

**Comment by andrewchang-bird** [comment]
Hi, Stijn, thanks for the detailed issue. If I’m understanding correctly, it looks like you’re trying to build Mockingbird from source using SPM, which might not be what you want. Instead, you should be able to add Mockingbird as a package dependency to your package manifest project and test target.

```swift
let package = Package(
  name: "MyPackage",
  dependencies: [
    .package(url: "https://github.com/birdrides/mockingbird.git", .upToNextMinor(from: "0.14.0")), // Add this line
  ],
  targets: [
    .testTarget(name: "MyPackageTests", dependencies: ["Mockingbird"]), // Add "Mockingbird" to dependencies
  ]
)
```

Unfortunately Mockingbird relies on an Xcode project to determine what sources and dependencies to parse. I’ll look into adding support for pulling that info from a `Package.swift` manifest file (e.g. using `swift package describe --type json`), but one workaround in the meantime is to simply generate a temporary backing Xcode project.

```bash
#!/bin/bash
set -eu

xcodeproj_path=MockingbirdSPMExample.xcodeproj

if [[ -d $xcodeproj_path ]]; then
  echo "Aborting because an Xcode project already exists at ${xcodeproj_path}"
  exit 1
fi

cleanup() {
  rm -rf $xcodeproj_path
}
trap cleanup EXIT

swift package generate-xcodeproj
mockingbird generate \
  --project "${xcodeproj_path}" \
  --target MockingbirdSPMExample \
  --output Tests/MockingbirdSPMExampleTests/MockingbirdSPMExampleMocks.generated.swift \
  --support Tests/MockingbirdSupport
```

You won’t get any of the benefits of an Xcode integration such as automatically generating the mocks before test runs or caching, but it should suffice for what you described.

Here is a complete example of the above: https://github.com/andrewchang-bird/mockingbird/tree/spm-package-example

**Comment by doozMen** [comment]
Oh super thanks for this detailed description. I will alter my use case that I was working on here https://github.com/doozMen/UseCaseInjectPropertyWrapperAndGeneratedMock

If it might help I wrote a mocking generator using Sourcery. Unfortunately I made to mistake of including it in a larger project called Highway. I learned a lot from Highway but the whole project is too big. I'm now distilling small things out of it. As I was extracting the generate mocks part I stumbled on your project which is a lot better at generating mocks. So thanks for releasing it! And if it might help here is the [sourcery on swift package](https://github.com/dooZdev/Highway/blob/master/Sources/HighwayLibrary/Commands/SourceryOnSwiftPackage.swift)

**Comment by doozMen** [comment]
Ok I tried out the example and noticed it states `// swift-tools-version:5.0`. In my case it was 5.2. When I try `swift package resolve` with 5.2 it outputs an error
```
'<#project#>' <#path#> error: dependency 'Mockingbird' in target '<#testTarget#>' requires explicit declaration; provide the name of the package dependency with '.package(name: "Mockingbird", url: "git@github.com:birdrides/mockingbird.git", from: "0.14.1")'
```
So maybe that is another issue related to swift 5.2?

**Comment by doozMen** [comment]
Another issue I have is related to the required platforms. I can understand for the cli tool that you need it to be a mac, but as my projects are server related I need it to support Linux as well. Could the platform requirements not be removed?

**Comment by andrewchang-bird** [comment]
If you add `.package(name: "Mockingbird", ...)` to the package dependency, SPM should allow you to use version `5.2`.

Note that the `platforms` definition in Mockingbird’s `Package.swift` manifest doesn’t affect Linux support and only constrains the minimum deployment version for that platform. (More info [here](https://docs.swift.org/package-manager/PackageDescription/PackageDescription.html#supportedplatform) and discussion [here](https://forums.swift.org/t/supportedplatform-linux-not-allowed/28500/2)).

**Comment by mygzi** [comment]
FYI `generate-xcodeproj` has been deprecated by the SPM team (see [here](https://github.com/apple/swift-package-manager/pull/3062)). 

> Starting Xcode 11, packages are directly supported by Xcode and the generate-xcodeproj command is no longer needed

It is also an awkward fit when using a local package (as you already have an enclosing xcodeproj to work with).

**Comment by doozMen** [comment]
Indeed would still be nice if this is supported out of the box. I know then we have to run mockingbird manually but this could maybe be done in `.git/hooks/pre-push` and on pull?

**Comment by andrewchang-bird** [comment]
Basic support via JSON project descriptions is available in 0.16 and compatibility with `swift package describe` is available in 0.17. See the wiki for more info: https://github.com/birdrides/mockingbird/wiki/Manual-Setup#generating-mocks-for-non-xcode-projects
